src/libre/re_comp.c

```diff
@@ -0,0 +1,320 @@
+			RECURSE(z, y, r);
+		} else {
+			NEWSTATE(z2);
+			RECURSE(x, z, l);
+			EPSILON(z, z2);
+			RECURSE(z2, y, r);
+		}
+
+		if (l->t == AST_EXPR_FLAGS) {
+			env->flags = l->u.flags.saved;
+		}
+		break;
+	}
+
+	case AST_EXPR_ALT:
+	{
+		struct fsm_state *la, *lz, *ra, *rz;
+		NEWSTATE(la);
+		NEWSTATE(lz);
+		NEWSTATE(ra);
+		NEWSTATE(rz);
+
+		EPSILON(x, la);
+		EPSILON(x, ra);
+		EPSILON(lz, y);
+		EPSILON(rz, y);
+
+		RECURSE(la, lz, n->u.alt.l);
+		RECURSE(ra, rz, n->u.alt.r);
+		break;
+	}
+
+	case AST_EXPR_LITERAL:
+		if (!addedge_literal(env, x, y, n->u.literal.c)) { return 0; }
+		break;
+
+	case AST_EXPR_ANY:
+		ANY(x, y);
+		break;
+
+	case AST_EXPR_MANY:
+		/* FIXME: is z necessary? */
+		NEWSTATE(z);
+		EPSILON(x, z);
+		ANY(x, z);
+		EPSILON(z, x);
+		EPSILON(z, y);
+		break;
+
+	case AST_EXPR_KLEENE:
+		EPSILON(x, y);
+		RECURSE(x, y, n->u.alt.l);
+		EPSILON(y, x);
+		break;
+
+	case AST_EXPR_PLUS:
+		RECURSE(x, y, n->u.alt.l);
+		EPSILON(y, x);
+		break;
+
+	case AST_EXPR_OPT:
+		RECURSE(x, y, n->u.alt.l);
+		EPSILON(x, y);
+		break;
+
+	case AST_EXPR_REPEATED:
+	{
+		struct fsm_state *na, *nz;
+		unsigned i, low, high;
+		struct fsm_state *a = NULL;
+		struct fsm_state *b = NULL;
+
+		/* This should have been normalized to [M;N], where
+		 * M < N and both are bounded, and {0,1} is built as
+		 * AST_EXPR_OPT instead. */
+		low = n->u.repeated.low;
+		high = n->u.repeated.high;
+		assert(low <= high);
+		assert(low != AST_COUNT_UNBOUNDED);
+		assert(high != AST_COUNT_UNBOUNDED);
+
+		/* Make new beginning/end states for the repeated section,
+		 * build its NFA, and link to its head. */
+		NEWSTATE(na);
+		NEWSTATE(nz);
+		RECURSE(na, nz, n->u.repeated.e);
+		EPSILON(x, na); /* link head to repeated NFA head */
+		b = nz;		/* set the initial tail */
+
+		if (low == 0) { EPSILON(na, nz); } /* can be skipped */
+
+		for (i = 1; i < high; i++) {
+			a = fsm_state_duplicatesubgraphx(env->fsm, na, &b);
+			if (a == NULL) { return 0; }
+
+			/* TODO: could elide this epsilon if fsm_state_duplicatesubgraphx()
+			 * took an extra parameter giving it a m->new for the start state */
+			EPSILON(nz, a);
+
+			/* To the optional part of the repeated count */
+			if (i >= low) { EPSILON(nz, b); }
+
+			na = a;	/* advance head for next duplication */
+			nz = b;	/* advance tail for concenation */
+		}
+
+		EPSILON(nz, y);	     /* tail to last repeated NFA tail */
+		break;
+	}
+
+	case AST_EXPR_CHAR_CLASS:
+	{
+		int i;
+		struct re_char_class_ast *cca = n->u.char_class.cca;
+		struct re_char_class *cc = re_char_class_ast_compile(cca);
+		if (cc == NULL) { return 0; }
+
+		for (i = 0; i < 256; i++) {
+			if (cc->chars[i/8] & (1U << (i & 0x07))) {
+				if (!addedge_literal(env, x, y, i)) { return 0; }
+			}
+		}
+		re_char_class_free(cc);
+		break;
+	}
+
+	case AST_EXPR_CHAR_TYPE:
+	{
+		int i;
+		struct re_char_class *cc = re_char_class_type_compile(n->u.char_type.id);
+		if (cc == NULL) { return 0; }
+
+		for (i = 0; i < 256; i++) {
```

Comment:
Any time you're iterating to 0..256, please make that `UCHAR_MAX` instead; it's something I grep for when considering character set support, and what can be bitmaps, and so on.